internal/beater/otlp/map_caching.go

```diff
@@ -0,0 +1,71 @@
+package otlp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/elastic/apm-server/internal/logs"
+	"github.com/elastic/apm-server/internal/sourcemap"
+	"github.com/elastic/elastic-agent-libs/logp"
+	lru "github.com/hashicorp/golang-lru"
+)
+
+type MapCachingFetcher struct {
+	cache   *lru.Cache
+	backend MapFetcher
+	logger  *logp.Logger
+}
+
+// NewMapCachingFetcher returns a MapCachingFetcher that wraps backend, caching results for the configured cacheExpiration.
+func NewMapCachingFetcher(
+	backend MapFetcher,
+	cacheSize int,
+) (*MapCachingFetcher, error) {
+	logger := logp.NewLogger(logs.Sourcemap)
+
+	lruCache, err := lru.NewWithEvict(cacheSize, func(key, value interface{}) {
```

Comment:
a little concerned about using an in-memory lru cache if the mapping files are going to be >10MB. We'll need to think about how much memory usage this will add. Otherwise, we'll have to cache some on local disk instead.

Reply:
Sure, what do you suggest we could do to verify whether this will use too much memory or not? Or should we just go ahead and cache it on the local disk right away? I'm not sure how to proceed here.

Reply:
I don't have the background knowledge here to make a decision on this. How large can a mapping file be, in normal cases and extreme cases? We'll have to ensure that `lru_size * mapping_file_size` is within reasonable memory usage while also reserving enough memory for normal APM server operation (e.g. aggregations). I'm highlighting this to make sure that we don't OOM just because we are caching large mapping files, especially for small APM servers (e.g. 1-2GB)

Reply:
Got it. Our opbeans app's map file size is about 7 MB, so considering that it's a dummy app, I'd say that it's quite easy for an app's map file size to get bigger than 10 MB. However, I don't think it's safe to say that it'd reach sizes bigger than 50 MB, so probably we can say that an extreme case could be around 50 MB (giving it some margin of error just to be safe). On top of that, it is expected that a single application can have multiple versions running at once, and each version must have its own map file. Therefore, I think it's very likely that a lot of map files can be in-memory at a time for a single app in case there are several versions of it being used (which tends to be a common scenario unless the app has some sort of force-upgrade mechanism built-in). I hope it helps to get a better idea of how much memory this feature can take.

Reply:
@simitt what do you think about this matter? Personally at least for smaller instances, the idea of caching a few 50MB mapping files sounds scary to me since it takes away a lot of memory headroom for apm-server. My feeling is that a local disk cache would make sense in this case.